[Tooling] Fix raw screenshots automation

AliSoftware opened the pull request:
Fixing various issues in the screenshot automation.

_See also [this update comment summarizing the changes](https://github.com/wordpress-mobile/WordPress-Android/pull/17096#issuecomment-1249942622) once I considered the PR ready for review._

## Raw Screenshots

 - [x] 👷  Completely reimplement the logic to launch emulators using Ruby and Fastlane
    - [x] Implement code to install the correct `system-image` depending on the API level we need + the current machine's architecture (Intel vs M1), using `sdkmanager`
    - [x] Re-implement the logic to create an AVD from a device, api level and system-image, using `avdmanager` instead of the `copy-device.sh` + `fixup-android-emulator.sh` shell scripts
    - [x] Clean up and get rid of the `fastlane/emulators/*.ini` and associated scripts
    - [x] Re-Implement the logic to kill previous emulators and launch a new `emulator` given an AVD, with the proper parameters
    - [x] Move the new code from `fastlane/helper/android_emulator_helper.rb` to the `release-toolkit` once it's confirmed to be stable --> https://github.com/wordpress-mobile/release-toolkit/pull/409
    - [x] Re-Implement the `screenshots` lane (taking the raw screenshots) and make it call the built-in fastlane action (`capture_android_screenshots`) directly now instead of our local `fastlane/action/take_android_emulator_screenshots.rb` one
 - [x] ⬆️ Update Screengrab library to latest version
 - [x] 🔧  Fix the `AndroidManifest.xml` permissions for debug
 - [x] 🕵️  Investigate and fix crash on `WPLocaleTestRule`
    - [x] 🐛  Fix NPE crash with the `Context` not yet initialized in the `AppInitializer.kt` singleton and `WordPress.kt` singleton by the time the test rule is applied — with the help of @ovitrif (see [those commits](https://github.com/wordpress-mobile/WordPress-Android/compare/fix/locale-override-in-ui-tests))
    - [x] 🔧  Align our custom implementation of `WPLocaleTestRule` with the new driving logic used by latest fastlane/screengrab version
    - [x] ⛓️  Use a `RuleChain` of fastlane's `LocaleTestRule` + our own `WPLocaleTestRule` to apply one on top of the other — and ensure fastlane's own internal logic in their `LocaleTestRule` implementation is called and not shadowed by our own rewriting of it as before
 - [x] 🕵️ 🐛   Investigate why passing `fr-FR` doesn't change the locale (but `fr` does).
    - According to fastlane's doc, we're supposed to pass `fr-FR`, `de-DE`, etc to the `locale` parameter, and that will end up being the name of the locale folder used in `fastlane/metadata/`.
    - But in practice it seems that Screengrab's `LocaleUtil.localeFromString` expects something else (is it an Android-like `de-rDE` / `zh-rCN` format? Or is it failing to find the locale for some other reason?). And if we pass a different locale code which works (like just `fr` instead of `fr-FR`) and changes the UI appropriately… then the `fastlane/*` locale folder that will be created will be incorrect…
    - ⚠️  This might be tricky to debug, since setting a `-e testLocale` Instrumentation Argument (from "Run/Debug Configurations" window) from Android Studio doesn't seem to work (the test locale that is read from the test and the `LocaleTestRule` returns `null` when running the test from Android Studio), and the passing of the `testLocale` only works when invoking the test using `am instrument -e testLocale …` from the command line (like fastlane's screengrab does)… but when running form the CLI we can't debug to inspect how the test locale is (incorrectly) parsed. And all the parsing methods that might be at fault are private to the Screengrab binary library, so that doesn't help debugging via some test code either.
 - [ ] 🕵️  Investigate why the JPScreenshotTest crashes on ZenDesk initialization (see below)
     -  ➡️  Tracked separately https://github.com/wordpress-mobile/WordPress-Android/issues/17130
 - [x] 🐛  Check the case of `zh-CN` which seems to prompt if we want to "Allow Google Pinyin input method" when the app starts, which interrupts the UI Test (see screenshot below). Can we use `adb shell` to force-disable this prompt?
 - [ ] 👷  Make the raw screenshots be saved in a separate temporary folder for WordPress vs Jetpack (see FIXMEs in 6eec4cfbbf5b155ae770956a176902cea481fbff)
     -  ➡️  Kept for a future refinement PR
 - [x] Test again the whole pipeline of running fastlane screenshots for multiple devices and multiple locales and ensure it switches from one to another as expected (i.e. that there's no leak across test run boundaries) and saves the PNGs in the expected folders

---

## Promo Screenshots

<details><summary>I will address those ones in a _separate_ PR, to keep that one focused only on fixing the raw screenshots.</summary>

 - [ ] Test the lane (not completely sure what's still broken vs left to be done since I didn't get that far last time given raw screenshots, which are an input for this, were still broken)
 - [ ] Update the `screenshots.json` to fix the device models and screenshot names being referenced
 - [ ] Update the device frames and background in `playstoreres` to match newer device models
 - [ ] Circle back with team Kāhu and Design team to update the compositing coordinates and config in `screenshots.json` with the new promo copies and their expected positioning/layout/rendering.
</details>

## Other

 - [ ] Update `SCREENSHOT_DEVICE_SETUP.md` docs ➡️  Will be done in dedicated PR
 - [x] Circle back with Annmarie and the Kāhu team to decide if it's expected that, given how the UI tests are written, some raw screenshots are asked to be taken in `light` mode and other screens switch to `dark` mode before doing the screenshot. ➡️  Tracked separately in https://github.com/wordpress-mobile/WordPress-Android/issues/17133

## Crash Log and issue details from items above (now solved)

<details><summary>ZenDesk Crash when trying to run the JPScreenshotTest on "phone" device (Pixel 3) in <tt>zh-CN</tt></summary>

```
$ /Users/olivierhalligon/Library/Android/sdk/platform-tools/adb -s emulator-5554 shell am instrument --no-window-animation -w \
-e testLocale zh-CN \
--no-hidden-api-checks \
-e appendTimestamp false \
-e class org.wordpress.android.ui.screenshots.JPScreenshotTest \
com.jetpack.android.test/org.wordpress.android.WordPressTestRunner
[19:58:54]: ▸ org.wordpress.android.ui.screenshots.JPScreenshotTest:
[19:58:54]: ▸ Process crashed while executing jPScreenshotTest(org.wordpress.android.ui.screenshots.JPScreenshotTest):
[19:58:54]: ▸ java.lang.IllegalArgumentException: Zendesk needs to be setup before this method can be called
[19:58:54]: ▸ 	at org.wordpress.android.support.ZendeskHelper.enablePushNotifications(ZendeskHelper.kt:197)
[19:58:54]: ▸ 	at org.wordpress.android.push.GCMRegistrationIntentService.sendRegistrationToken(GCMRegistrationIntentService.java:86)
[19:58:54]: ▸ 	at org.wordpress.android.push.GCMRegistrationIntentService.lambda$onHandleWork$0$org-wordpress-android-push-GCMRegistrationIntentService(GCMRegistrationIntentService.java:53)
[19:58:54]: ▸ 	at org.wordpress.android.push.GCMRegistrationIntentService$$ExternalSyntheticLambda0.onComplete(Unknown Source:2)
[19:58:54]: ▸ 	at com.google.android.gms.tasks.zzi.run(com.google.android.gms:play-services-tasks@@18.0.1:1)
[19:58:54]: ▸ 	at android.os.Handler.handleCallback(Handler.java:873)
[19:58:54]: ▸ 	at android.os.Handler.dispatchMessage(Handler.java:99)
[19:58:54]: ▸ 	at android.os.Looper.loop(Looper.java:193)
[19:58:54]: ▸ 	at android.app.ActivityThread.main(ActivityThread.java:6669)
[19:58:54]: ▸ 	at java.lang.reflect.Method.invoke(Native Method)
[19:58:54]: ▸ 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
[19:58:54]: ▸ 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
[19:59:09]: ▸ INSTRUMENTATION_RESULT: shortMsg=Process crashed.
```
</details>

<details><summary> 🖼️ Running Screenshots in <tt>zh-CN</tt> interrupts the UI tests with an alert from the system asking if we want to enable Google Pinyin input method</summary>

<img width="1567" alt="image" src="https://user-images.githubusercontent.com/216089/186971685-e6c7bf8c-33a4-491a-8aa1-82316d1a5737.png" />
<img width="868" alt="image" src="https://user-images.githubusercontent.com/216089/186975691-a1243e11-6e98-48ee-905c-a7c371424815.png" />

</details>

---

## To Test

ℹ️  See https://github.com/wordpress-mobile/release-toolkit/pull/409 for a more detailed set of instructions for testing this.

 - Run `bundle exec fastlane screenshots app:wordpress`, go take a very long walk while it runs, and check that all the expected raw screenshots are generated in `fastlane/screenshots/raw` for all locales and both device types
 - Run `bundle exec fastlane screenshots app:jetpack`, go take a very long walk while it runs, and check that all the expected raw screenshots are generated in `fastlane/screenshots/raw` for all locales and both device types

Optionally, during iterative testing, you can limit the runs to a subset of devices and locales, e.g. `bundle exec fastlane screenshots app:wordpress device:phone locale:ar,fr-FR,zh-CN` to avoid running it for all Mag16 and both devices for each app.

Though ultimately you will need to test all locales and both devices to be sure that:
 - The device codes are correct and thus the screenshots show the UI in the expected language for each locale (instead of using English, or using the locale from the previous run of the emulator…)
 - The device models used are as expected, and that those devices exist and can have their `system-image` be downloaded, their AVD be created, their emulator launched, and the permissions needed by fastlane/screengrab are set on those devices without any error (e.g. without them needing root access, or failing to get a permission for that emulator model and system image…)

AliSoftware committed:
WIP: screenshots lane

AliSoftware committed:
Update `tools.fastlane:screengrab` dependency to latest

AliSoftware committed:
WIP: Start AVD before triggering screenshots

AliSoftware committed:
WIP: rewrite emulator helper

AliSoftware committed:
Extract `install_system_image` in its own method

AliSoftware committed:
Do a cold boot of emulators

AliSoftware committed:
Add docs about why we don't use `reboot -p`

AliSoftware committed:
Update debug AndroidManifest.xml permissions

AliSoftware committed:
Call calling capture_android_screenshots for real

AliSoftware committed:
Fix issue about frozen variables in lane

AliSoftware committed:
Fix inconsistency in locales list

Entry for `is` did not have `promo_config: false` so it was selected as part of `SCREENSHOT_LOCALES`, and this despite not having a value for `google_play`

AliSoftware committed:
Provide `*_package_name` parameters explicitly

AliSoftware committed:
Fix Manifest permissions

ovitrif committed:
Make sure AppInitializer context is only set internally

ovitrif committed:
Use context from testApp when setting locale in WPLocaleTestRule, since it runs before AppInitalizer inits

ovitrif committed:
Cleanup context ref in updateContextLocale

ovitrif committed:
Put @JvmOverloads on separate line

AliSoftware committed:
Update WPLocaleTestRule + chain it with fastlane's

The latest version of fastlane Screengrab lib does not pass the `endingLocale` when triggering the instrumentation tests, and that fastlane's own built-in `LocaleTestRule` implementation has changed to their latest way to do it, so we need to update our custom `WPLocaleTestRule` (which is based on the model of fastlane's `LocaleTestRule`) accordingly.

Then, to ensure that both the code from fastlane's own `LocaleTestRule` (which consists of changing the device locale but also calling `Screengrab.setLocale` and setting internal stuff for screengrab to work properly and as expected) and our own code from `WPLocaleTestRule` (which handles our WP-specific, in-app locale switching logic) are run and are executed in the right order (to ensure that both fastlane's rule and expectations are respected and that our custom logic is handled), we need to use a `RuleChain` on each of our screenshot tests (`{WP,JP}ScreenshotTest.java`)

---

Note that if I'm not mistaken, our `WPLocaleTestRule` is not really necessary when we're in the case where we run the screenshot tests using `fastlane screenshots` lane and start on a fresh and wiped-out emulator where the APK have never been installed, as in such case the in-app locale switcher would never have been changed by a user, and the User Preference for the custom locale would be `null`, and the app will then always use the Device Locale (which fastlane's `LocaleTestRule` takes care of changing) in such case. But I still think it's valuable to add our `WPLocaleTestRule` to the `RuleChain` for the case when developers work on fixing the screenshot UI tests and might try to run them from Android Studio, via an emulator they might not have wiped out first, and on which they might already have the APK installed from previous work, and in which they might have changed the in-app locale. So better safe than sorry :)

AliSoftware committed:
Allow to restrict the run to more than one locale

AliSoftware committed:
Use `-wipe-data` on emulators by default

AliSoftware committed:
Fix test class for Jetpack screenshots

AliSoftware committed:
🚧 Add note to remember to make screenshots/ dirs app-specific

Easy enough for the raw screenshots and the `:screenshots` lane, but we should make the change at the same time that we do it for the `:promo_screenshots` too

AliSoftware committed:
Clean up old and obsolete scripts

AliSoftware committed:
Fix checkstyle violation

AliSoftware committed:
Fix tool lookup logic on emulator helper

AliSoftware commented on fastlane/helpers/android_emulator_helper.rb:
```diff
@@ -0,0 +1,190 @@
+      end
+
+      def sdkmanager
+        @sdkmanager_bin ||= tool(paths: ['cmdline-tools', 'latest', 'bin'], binary: 'sdkmanager')
```
TODO: I think depending on the installation and tool version and machine, sometimes it's called something other than `latest` (like `current` or something? At least I vaguely remember having to deal with that difference between my local setup and how it was called on CI in a previous PR on another repo, even if I can't remember exactly where)

AliSoftware commented on fastlane/helpers/android_emulator_helper.rb:
```diff
@@ -0,0 +1,190 @@
+        end
+      end
+
+      def retry_loop(time_between_retries:, timeout:, description:)
```
TODO: Add YARD documentation for this method

AliSoftware commented on fastlane/lanes/screenshots.rb:
```diff
@@ -1,126 +1,86 @@
+# FIXME: Make this app dependent instead (WordPress vs Jetpack)
 RAW_SCREENSHOTS_DIR = File.join(Dir.pwd, "screenshots", "raw")
 RAW_SCREENSHOTS_PROCESSING_DIR = File.join(Dir.pwd, "screenshots", "raw_tmp")
 PROMO_SCREENSHOTS_PROCESSING_DIR = File.join(Dir.pwd, "screenshots", "promo_tmp")
-FINAL_METADATA_DIR = File.join(Dir.pwd, "metadata/android")
-
+FINAL_METADATA_DIR = File.join(Dir.pwd, "metadata", "android")
```
Reminder to fix this (i.e. use the proper `{jetpack_,}metadata` subfolder depending on WP vs JP app) before marking this PR as ready

AliSoftware replied:
Note: This will be done as part of a subsequent PR after all (see `tooling/screenshots-lane-tweaks` branch) to not hold this one any longer.

AliSoftware commented on fastlane/lanes/screenshots.rb:
```diff
@@ -1,126 +1,86 @@
+  .compact
+  .freeze
+
+require_relative '../helpers/android_emulator_helper'
```
TODO: Move this to release-toolkit before considering this PR ready

AliSoftware commented on fastlane/lanes/screenshots.rb:
```diff
@@ -143,7 +103,7 @@
     locales = ALL_LOCALES
       .select { |hsh| hsh[:promo_config] != false }
-      .map {| hsh | [ hsh[:glotpress], hsh[:google_play] ]}
+      .map { |h| h[:google_play] }
```
⚠️  This might be a copy/paste error, as it seems it was returning an array of `[glotpress, googleplay]` items before but is now using a single one.

I think that's when I tried to fix the incorrect list of locales and then copy/pasted the fix everywhere for consistency and missed the subtle difference here.

(Can't wait to have time to finish my work on the `LocaleHelper` in the `release-toolkit` at some point in the future to have all this locale manipulation simplified!)

AliSoftware replied:
Fixed in 1dbc2c995a7c85f64bef883054f3fa77a046c08c

AliSoftware commented on fastlane/lanes/screenshots.rb:
```diff
@@ -302,30 +261,44 @@
+  desc "Rebuild emulators used for screenshots"
+  lane :create_emulators do |options|
+    SCREENSHOT_DEVICES.each do |device|
+      create_avd(api: device[:api], device: device[:device])
```
TODO: We need to call `install_system_image(api: device[:api])` first here.

Though in practice we might as well get rid of that lane, now that I've refactored the `screenshots` lane to directly call `create_avd` on its own loop — in order to only create the emulators for which screenshots are requested to be build, instead of all of them even the ones we filtered out — instead of having the `screenshots` lane call `create_emulators` at its start like before.

AliSoftware replied:
Fixed in d418d86d3c67b1e0355c98b5f3735b4dea07ae4f

AliSoftware committed:
WIP: Check locale codes expected

ovitrif commented on WordPress/src/androidTest/java/org/wordpress/android/ui/screenshots/WPLocaleTestRule.java:
```diff
@@ -1,57 +1,53 @@
+        Context context = ApplicationProvider.getApplicationContext();
+        String original = LocaleManager.getLanguage(context);
+        LocaleManager.setNewLocale(context, localeCode);
+        WordPress.updateContextLocale(ApplicationProvider.getApplicationContext());
```
A small np if I may: same `context` captured in the variable can be passed here to DRY the code 👍

AliSoftware replied:
🤦 Good catch… especially since I initially created the variable exactly to DRY the code in the intention to use it for all three calls… not sure why I missed the last one 😅

AliSoftware committed:
DRY the `context` variable in `WPLocaleTestRule`

Kudos to @ovitrif for spotting this in https://github.com/wordpress-mobile/WordPress-Android/pull/17096#discussion_r957525144

AliSoftware committed:
Fix Locale expecting different formats

The invocation of the Instrumented Test by fastlane / screengrab expects a locale code like `fr-FR` (which is what it will use as the name, while the Android runtime expects a locale code like `fr_FR` to be able to parse it and interpret it as a valid `java.utils.Locale`

AliSoftware committed:
Try disabling Pinyin IME before testing in zh-CN

AliSoftware committed:
Revert "Try disabling Pinyin IME before testing in zh-CN"

This reverts commit 1bcce5091466f5240086aa74a3ecb8e159d75cdf.

AliSoftware committed:
Force IME to Latin using a TestRule

AliSoftware committed:
Make the ImeTestRule a good citizen

By restoring the previous IME when it finishes

AliSoftware committed:
Replicate Rules to JPScreenshotTest as well

AliSoftware committed:
Avoid NPE when running from AS

When running the test from AndroidStudio, the `testLocale` parameter is not passed to the InstrumentationRegistry so the `testLocale` is `null`

AliSoftware committed:
Make the ImeTestRule embed the GrandPermissionRule

So that at call site we only need to declare the ImeTestRule in the test, without having to know about the implementation detail of it having to be chained with a GrantPermissionRule first.

AliSoftware committed:
🚧 Move android_emulator_helper to release-toolkit

This points to the branch of the open PR, until it lands and a new version gets released.

AjeshRPai committed:
Fixes: Inconsistency of Light mode and Dark mode in screenshots

Updates the logic to make all the screenshots to be in light mode.

AjeshRPai committed:
Moves: Switching to light mode logic to the parent function

AliSoftware commented:
## Update: Ready for Review

All critical/blocking tooling bugs around the `screenshots` lane (raw screenshots) have been solved now. So I think that PR is ready to review and land now 🎉 

> **Warning**: This PR still depends on https://github.com/wordpress-mobile/release-toolkit/pull/409 so we will need to land that one in `release-toolkit`, then update the `fastlane/Pluginfile` after we created a new toolkit release, before we can merge this one here.

---

### Remaining, non-blocking tooling refinements for a future PR

Those will be done in separate PRs.

  - Make `fastlane screenshots app:<app>` generate the raw screenshots in different folders for WordPress vs Jetpack
  - See if we can avoid the red messages that appear in the logs when fastlane tries to delete old screenshots for the device's sdcard and can't find any. This will likely need to be fixed [inside `fastlane`'s code itself](https://github.com/fastlane/fastlane/blob/5eaf30d538f6df225f743cbb773346c47a7bcdd3/screengrab/lib/screengrab/runner.rb#L168-L181)
  - Update `SCREENSHOT_DEVICE_SETUP.md` docs
  - Test, and fix if needed, the `promo_screenshots` lane now that we have working raw screenshots that we can use as input

### Remaining non-tooling fixes (UI test failures in Android) — fixed separately

The are also things that will need fixing, but on the Android code of the UI tests themselves (ZenDesk crash, Hilt flaky crash, light vs dark mode…).

 - Those have been gathered under [this parent GitHub issue](https://github.com/wordpress-mobile/WordPress-Android/issues/17128)
 - All those are not tooling-related, but Android-code-related (UI test not passing); so they will be fixed separately. 
 - I made a call for help internally [ref: pcdRpT-MB-p2] and @AjeshRPai responded to the call; he's currently already working on some of those (for example https://github.com/wordpress-mobile/WordPress-Android/pull/17166)

AjeshRPai committed:
Moves: Light mode logic to parent function

Earlier the logic to set light mode was called from each function on screenshot. This has been moved to the parent function since all the screenshots needed are on Light mode.

AjeshRPai committed:
Merge pull request #17166 from wordpress-mobile/issue/17133-screenshots-inconsistency-between-light-and-dark-mode

Fixes: Inconsistency of Light mode and Dark mode in screenshots

ParaskP7 commented on WordPress/src/androidTest/java/org/wordpress/android/ui/screenshots/WPLocaleTestRule.java:
```diff
@@ -1,57 +1,61 @@
+
+import tools.fastlane.screengrab.locale.LocaleUtil;
 
+public class WPLocaleTestRule implements TestRule {
```
Question (❓): The below was copied from the commit message itself, that is, part of the description on it:

```
But I still think it's valuable to add our `WPLocaleTestRule` to the `RuleChain`
for the case when developers work on fixing the screenshot UI tests and might
try to run them from Android Studio, via an emulator they might not have
wiped out first, and on which they might already have the APK installed from
previous work, and in which they might have changed the in-app locale.
So better safe than sorry :)
```

I want to challenge this a bit in favor of removing our custom `WPLocaleTestRule` no matter. However, FYI, I have no strong opinions anyway, I just want to challenge this in favor of ending-up with less code/configuration to maintain for later on.

As such, IMHO, for the case when engineers work on fixing the screenshots UI tests and might run them from AS, I think it is better to document that case and make it explicit, so that engineers would always wipe out the emulator first or uninstall the APK first. I think it is better to have a strict and documented process on something that would rarely be the case, than having a custom rule helping us overcome it, in cases where engineers will need it, just because they missed to read the clear instructions on how to use the screenshot solution. Wdyt?

PS: By removing the `WPLocaleTestRule`, apart from the obvious code/configuration wins, we can get rid of the `WordPress.updateContextLocale(...)` in favor of using the `AppInitializer.updateContextLocale()` direct within the `AppSettingsFragment.changeLanguage(...)` method, it being the only usage afterwards, as well as reverting it to become as it was, that is, without it accepting the extra `appContext` parameters.

AliSoftware replied:
That's an interesting take.

So basically, remove the `WPLocaleTestRule`, and add a comment in the code of the test classes to tell developers to ensure they uninstall any previous version of the app before testing to ensure that the in-app locale switched and user preference for overridden locale does not accidentally take over the device locale set by the `LocaleTestRule` from fastlane?

I'm ok with that tbh. Basically that fact of keeping `WPLocaleTestRule` came to be because the code and class for the rule was already there before I started this PR, so I didn't think about removing it and instead tried to fix it. But if it can help remove complexity and our dance to have to fix the NPE for the `AppInitializer.kt` singleton being initialized too late etc… then maybe that's simpler indeed.

And once we'll migrate to Android 13 and the new way of doing per-app locale switching and get rid of all our logic for our custom in-app locale switching code that is around all that… we probably won't need any of this `WPLocaleTestRule` (since `WordPress.updateContextLocale(…)` and `AppInitializer.updateContextLocale()` will both have disappeared and rendered obsolete by us migrating to Android 13 and letting the OS handle that logic for us, right?)

ParaskP7 replied:
> So basically, remove the WPLocaleTestRule, and add a comment in the code of the test classes to tell developers to ensure they uninstall any previous version of the app before testing to ensure that the in-app locale switched and user preference for overridden locale does not accidentally take over the device locale set by the LocaleTestRule from fastlane?

Exactly! 💯 

> I'm ok with that tbh. Basically that fact of keeping WPLocaleTestRule came to be because the code and class for the rule was already there before I started this PR, so I didn't think about removing it and instead tried to fix it.

Yeap, I figured as much, I think it might be wise to take this chance and remove it now that we can (avoid having it there just for this case). Also, with this, we would potentially avoid the risk of adding to it later, for some unknown reason, thus risking making it indispensable and even harder to remove afterwards.

> But if it can help remove complexity and our dance to have to fix the NPE for the AppInitializer.kt singleton being initialized too late etc… then maybe that's simpler indeed.

Exactly! 💯 

> And once we'll migrate to Android 13 and the new way of doing per-app locale switching and get rid of all our logic for our custom in-app locale switching code that is around all that… we probably won't need any of this WPLocaleTestRule (since WordPress.updateContextLocale(…) and AppInitializer.updateContextLocale() will both have disappeared and rendered obsolete by us migrating to Android 13 and letting the OS handle that logic for us, right?)

True, but again, I think this might be a slight overstatement as us migrating to Android 13 doesn't mean that all of the devices are going to be utilizing Android 13. We would still need to support Android 12 and below for years to come. 🤔

However, if you mean that we can then update our emulators from our current usage of `API 28 (Android 9)` to using the newest `API 33 (Android 13)` for the screenshot generation, then yes, this is doable. 👍

AliSoftware replied:
> However, if you mean that we can then update our emulators from our current usage of API 28 (Android 9) to using the newest API 33 (Android 13) for the screenshot generation, then yes, this is doable. 👍

Yep, that's what I had in mind.

For the record, we will very likely update the API level of the emulators used by the screenshot automation in a separate PR, from API 28 to at least API 31… if not 33, as part of fixing the "Gray overlay sometimes appears on views" bug mentioned by @zwarm in https://github.com/wordpress-mobile/WordPress-Android/issues/16900 that seems to only happen on API<30.

I just haven't updated those as part of this PR because that PR focused on making the automation work again in its current state, so preferred to keep our exploration of that gray overlay and API level bump in separate future PR; but if we update the API level of those emulators to API 33 as part of this future PR, that would be one more reason for us to get rid of `WPLocaleTestRule` in this current PR, considering that we hope to then soon (-ish 🤞) remove our in-app locale switcher on devices that are using Android 13 anyway.

ParaskP7 replied:
Got it, thanks so much for the clarification on that Olivier! 👍 

> ...from API 28 to at least API 31… if not 33

I would recommend going with `API 31` for now as we are not ready for `API 33` just yet and I believe, we will not be there for some time...

ParaskP7 commented on WordPress/src/debug/AndroidManifest.xml:
```diff
@@ -27,4 +27,15 @@
+    <!-- Allows changing SystemUI demo mode for screenshot automation -->
+    <uses-permission
+            android:name="android.permission.DUMP"
+            tools:ignore="ProtectedPermissions" />
```
1. Minor (🔍): The configuration above can be automatically reformatted (see `Reformat Code` action):
```
    <!-- Allows forcing IME for screenshot automation -->
    <uses-permission
        android:name="android.permission.WRITE_SECURE_SETTINGS"
        tools:ignore="ProtectedPermissions" />

    <!-- Allows changing SystemUI demo mode for screenshot automation -->
    <uses-permission
        android:name="android.permission.DUMP"
        tools:ignore="ProtectedPermissions" />
```
2. Minor (🔍): Actually, it seems the whole file can be automatically reformatted as there are some other minor formatting adjustments that can be fixed.
3. Minor (🔍): It is recommended that the `<application ... />` node sits below the `<uses-permission ... />` nodes, actually AS is complaining about that. Also, the `application` node can be reformatted as well:
```

    <application
        android:name=".WordPressDebug"
        android:supportsRtl="true"
        tools:replace="android:name,android:supportsRtl" />
```

AliSoftware replied:
Fixed in 238d880f421647f4af06c469ef015f5005404c96

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/AppInitializer.kt:
```diff
@@ -890,6 +890,7 @@ class AppInitializer @Inject constructor(
         private const val DEFAULT_TIMEOUT = 2 * 60 // 2 minutes
 
         @SuppressLint("StaticFieldLeak") var context: Context? = null
+            private set;
```
Minor (🔍): This is a Kotlin file and as such consider omitting the trailing `;` at the end of the statement.

AliSoftware replied:
Fixed in bae3f3ee179ddd234a3b7bbcf4c18eb9908a9f28

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/AppInitializer.kt:
```diff
@@ -1019,8 +1020,9 @@ class AppInitializer @Inject constructor(
-        fun updateContextLocale() {
-            context = LocaleManager.setLocale(context)
+        fun updateContextLocale(appContext: Context? = null) {
+            val context = appContext ?: context!!
```
Question (❓): Why were the double exclamation marks (`!!`) added to the `context`? Is that by design, do we want the `updateContextLocale(...)` function to crash exactly there, that is, when the `context` is not already initialized in advance? 🤔

AliSoftware replied:
tbh I have no clue about this one, you'll have to ask @ovitrif as the author of this commit 😅 

The reason for this change on `updateContextLocale` is that when we tried to run the UI tests before that change, the `TestRules` that we have on the screenshot tests tried to change the context before the `AppInitializer.kt` singleton was initialized (as the `WPLocaleTestRule` is a `@ClassRule` and executed way before all that) which led to a NPE (see the first checkbox with a 🕵️ in my PR description), and which is the reason why we had to change the code to pass the context to `updateContextLocale` here.

But know being a Kotlin dev, I have no idea if that `!!` vs `!` vs other solutions was the right one for this implementation of the fix.

ovitrif replied:
> Question (❓): Why were the double exclamation marks (!!) added to the context? Is that by design, do **we want the `updateContextLocale(...)` function to crash exactly there, that is, when the context is not already initialized in advance?** 🤔

Good thinking @ParaskP7!

Initially investigation into the issue described by @AliSoftware started with an error thrown somewhere deep in the code when a "random" (not easily leading to the root cause) method was getting called, the only relevant clue we got was that the `context` was null but this wasn't very immediate eiter.

To improve things, I updated the code to fail fast, and as a result of more refactoring I've ended up with this solution to make sure we're avoiding the need to check too deep in the code in this case.

In principle, it was my way of telling "you can't call this method" if you don't provide the `ApplicationContext` unless it was already initialized.

> But know being a Kotlin dev, I have no idea if that !! vs ! vs other solutions was the right one for this implementation of the fix.

I've been reasoning about `!` from Swift sometimes associating it with the `!!` from Kotlin 😄 , it could be seen as such in some situations :)

ParaskP7 replied:
Oh cool, thanks so much for the extra clarification on that @ovitrif ! 🙇 

Can I maybe suggest a follow-up on that then, maybe adding a comment or some kind of documentation on that, that is, in case we anyway proceed with keeping the `WPLocaleTestRule` test rule. I recommend this action just so that to make sure we avoid this kind of question in the future, or avoid having someone like me, when looking as that, to be thinking, oh, but this `!!` is unnecessary, and ending up removing it silently anyway. Wdyt? 🤔

ovitrif replied:
Sounds good to me Petros 👍

AliSoftware replied:
@ParaskP7 While I mainly agree that maybe we should remove `WPLocaleTestRule` after all (as discussed in the other comment thread of this PR), what will likely happen is that I'll keep it in this PR for now because I want to avoid making another big change on this PR (like removing `WPLocaleTestRule`) if possible so that it doesn't risk delaying the landing of that PR (since more follow-up work, including some work to fix Android issues like #17128, depend on this to land, so I don't want to block those either).

So my plan is:
 - Let @ovitrif  address your suggestion on adding the adequate code comment on the `!!`
 - Hopefully find someone with a M1 (and ideally someone who doesn't work on Android daily and thus might not have the necessary tools and system images installed on their Mac, to ensure the lane errors correctly and installs the necessary things when they are not installed yet…) to add an additional review to this PR… though if no one manifests I'll likely end up merging it anyway to move forward
 - Consider removing `WPLocaleTestRule` in a follow-up PR. Which means at this point the changes around `updateContextLocale` now taking `context` as a parameter instead of infering it from local value would likely stay even when I'll remove `WPLocaleTestRule`… and I think that's not a bad thing (Dependency Injection over Singleton and inner hardcoded values ftw) if we keep those changes made by Ovi anyway even if the TestRule disappears?

ParaskP7 replied:
Sounds like a plan to me, thanks for the extra clarification on keep the `WPLocaleTestRule` @AliSoftware , much appreciated! 🥇

> ...and I think that's not a bad thing if we keep those changes made by Ovi anyway even if the TestRule disappears?

I am not sure on that, but I don't have a strong opinion on it too. I am always in favor of less complexity, less code, less code related documentation (when possible). Thus, if you feel that even after removing this `WPLocaleTestRule` we might still benefit from all that, then I wouldn't debate on it again. All I wanted to do is becoming the devil's advocate and make you double think this update. 😈

ovitrif replied:
Added method docs + contract check and error message via this suggestion PR: https://github.com/wordpress-mobile/WordPress-Android/pull/17196 👍

ParaskP7 reviewed (commented):
👋 @AliSoftware !

I have reviewed and tested this PR as per the description on both, this and its [accompanying PR](https://github.com/wordpress-mobile/release-toolkit/pull/409), everything worked (almost) as expected, good job in re-doing the `screenshots` lane! 🥇 

This (almost) mention above, is due to the fact that although running a command like `bundle exec fastlane screenshots app:wordpress device:phone locale:de-DE,fr-FR` for `WordPress` worked as expected, it didn't work for `Jetpack`, that is, when I replaced the `app:wordpress` with `app:jetpack`, is that expected? See crash below:

<details>
  <summary>Click me</summary>

```
...
[15:32:17]: ▸ org.wordpress.android.ui.screenshots.JPScreenshotTest:
[15:32:17]: ▸ Process crashed while executing jPScreenshotTest(org.wordpress.android.ui.screenshots.JPScreenshotTest):
[15:32:17]: ▸ java.lang.IllegalStateException: The component was not created. Check that you have added the HiltAndroidRule.
[15:32:17]: ▸ 	at dagger.hilt.internal.Preconditions.checkState(Preconditions.java:83)
[15:32:17]: ▸ 	at dagger.hilt.android.internal.testing.TestApplicationComponentManager.generatedComponent(TestApplicationComponentManager.java:96)
[15:32:17]: ▸ 	at org.wordpress.android.WordPressTest_Application.generatedComponent(WordPressTest_Application.java:28)
[15:32:17]: ▸ 	at dagger.hilt.EntryPoints.get(EntryPoints.java:59)
[15:32:17]: ▸ 	at org.wordpress.android.WordPress.component(WordPress.kt:26)
[15:32:17]: ▸ 	at org.wordpress.android.push.GCMMessageService.onCreate(GCMMessageService.java:42)
[15:32:17]: ▸ 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3532)
[15:32:17]: ▸ 	at android.app.ActivityThread.access$1300(ActivityThread.java:199)
[15:32:17]: ▸ 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1666)
[15:32:17]: ▸ 	at android.os.Handler.dispatchMessage(Handler.java:106)
[15:32:17]: ▸ 	at android.os.Looper.loop(Looper.java:193)
[15:32:17]: ▸ 	at android.app.ActivityThread.main(ActivityThread.java:6669)
[15:32:17]: ▸ 	at java.lang.reflect.Method.invoke(Native Method)
[15:32:17]: ▸ 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
[15:32:17]: ▸ 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
[15:32:17]: ▸ Process crashed while executing jPScreenshotTest(org.wordpress.android.ui.screenshots.JPScreenshotTest):
[15:32:17]: ▸ java.lang.RuntimeException: Unable to create service org.wordpress.android.push.GCMMessageService: java.lang.IllegalStateException: The component was not created. Check that you have added the HiltAndroidRule.
[15:32:17]: ▸ 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3544)
[15:32:17]: ▸ 	at android.app.ActivityThread.access$1300(ActivityThread.java:199)
[15:32:17]: ▸ 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1666)
[15:32:17]: ▸ 	at android.os.Handler.dispatchMessage(Handler.java:106)
[15:32:17]: ▸ 	at android.os.Looper.loop(Looper.java:193)
[15:32:17]: ▸ 	at android.app.ActivityThread.main(ActivityThread.java:6669)
[15:32:17]: ▸ 	at java.lang.reflect.Method.invoke(Native Method)
[15:32:17]: ▸ 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
[15:32:17]: ▸ 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
[15:32:17]: ▸ Caused by: java.lang.IllegalStateException: The component was not created. Check that you have added the HiltAndroidRule.
[15:32:17]: ▸ 	at dagger.hilt.internal.Preconditions.checkState(Preconditions.java:83)
[15:32:17]: ▸ 	at dagger.hilt.android.internal.testing.TestApplicationComponentManager.generatedComponent(TestApplicationComponentManager.java:96)
[15:32:17]: ▸ 	at org.wordpress.android.WordPressTest_Application.generatedComponent(WordPressTest_Application.java:28)
[15:32:17]: ▸ 	at dagger.hilt.EntryPoints.get(EntryPoints.java:59)
[15:32:17]: ▸ 	at org.wordpress.android.WordPress.component(WordPress.kt:26)
[15:32:17]: ▸ 	at org.wordpress.android.push.GCMMessageService.onCreate(GCMMessageService.java:42)
[15:32:17]: ▸ 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3532)
[15:32:17]: ▸ 	... 8 more
[15:32:17]: ▸ INSTRUMENTATION_RESULT: shortMsg=Process crashed.
[15:32:17]: ▸ INSTRUMENTATION_CODE: 0
...
[15:35:40]: ▸ org.wordpress.android.ui.screenshots.JPScreenshotTest:
[15:35:40]: ▸ Error in jPScreenshotTest(org.wordpress.android.ui.screenshots.JPScreenshotTest):
[15:35:40]: ▸ androidx.test.espresso.base.DefaultFailureHandler$AssertionFailedWithCauseError: 'an instance of android.widget.TextView and view.getText() with or without transformation to match: is "Choose site"' doesn't match the selected view.
[15:35:40]: ▸ Expected: an instance of android.widget.TextView and view.getText() with or without transformation to match: is "Choose site"
[15:35:40]: ▸      Got: view.getText() was "Choisir un site" transformed text was "Choisir un site"
[15:35:40]: ▸ View Details: AppCompatTextView{id=-1, visibility=VISIBLE, width=368, height=74, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.appcompat.widget.Toolbar$LayoutParams@ab38b50, tag=null, root-is-layout-requested=false, has-input-connection=false, x=198.0, y=40.0, text=Choisir un site, input-type=0, ime-target=false, has-links=false}
[15:35:40]: ▸ 	at dalvik.system.VMStack.getThreadStackTrace(Native Method)
[15:35:40]: ▸ 	at java.lang.Thread.getStackTrace(Thread.java:1538)
[15:35:40]: ▸ 	at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:12)
[15:35:40]: ▸ 	at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:7)
[15:35:40]: ▸ 	at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:5)
[15:35:40]: ▸ 	at androidx.test.espresso.ViewInteraction.check(ViewInteraction.java:12)
[15:35:40]: ▸ 	at org.wordpress.android.e2e.pages.MySitesPage.switchSite(MySitesPage.java:62)
[15:35:40]: ▸ 	at org.wordpress.android.e2e.pages.MySitesPage.startNewSite(MySitesPage.java:85)
[15:35:40]: ▸ 	at org.wordpress.android.ui.screenshots.JPScreenshotTest.generateSiteTopic(JPScreenshotTest.java:213)
[15:35:40]: ▸ 	at org.wordpress.android.ui.screenshots.JPScreenshotTest.jPScreenshotTest(JPScreenshotTest.java:135)
[15:35:40]: ▸ 	... 41 trimmed
[15:35:40]: ▸ Caused by: junit.framework.AssertionFailedError: 'an instance of android.widget.TextView and view.getText() with or without transformation to match: is "Choose site"' doesn't match the selected view.
[15:35:40]: ▸ Expected: an instance of android.widget.TextView and view.getText() with or without transformation to match: is "Choose site"
[15:35:40]: ▸      Got: view.getText() was "Choisir un site" transformed text was "Choisir un site"
[15:35:40]: ▸ View Details: AppCompatTextView{id=-1, visibility=VISIBLE, width=368, height=74, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.appcompat.widget.Toolbar$LayoutParams@ab38b50, tag=null, root-is-layout-requested=false, has-input-connection=false, x=198.0, y=40.0, text=Choisir un site, input-type=0, ime-target=false, has-links=false}
[15:35:40]: ▸ 	at androidx.test.espresso.matcher.ViewMatchers.assertThat(ViewMatchers.java:16)
[15:35:40]: ▸ 	at androidx.test.espresso.assertion.ViewAssertions$MatchesViewAssertion.check(ViewAssertions.java:11)
[15:35:40]: ▸ 	at androidx.test.espresso.ViewInteraction$SingleExecutionViewAssertion.check(ViewInteraction.java:2)
[15:35:40]: ▸ 	at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:12)
[15:35:40]: ▸ 	at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:1)
[15:35:40]: ▸ 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[15:35:40]: ▸ 	at android.os.Handler.handleCallback(Handler.java:873)
[15:35:40]: ▸ 	at android.os.Handler.dispatchMessage(Handler.java:99)
[15:35:40]: ▸ 	at android.os.Looper.loop(Looper.java:193)
[15:35:40]: ▸ 	at android.app.ActivityThread.main(ActivityThread.java:6669)
[15:35:40]: ▸ 	at java.lang.reflect.Method.invoke(Native Method)
[15:35:40]: ▸ 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
[15:35:40]: ▸ 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
[15:35:47]: ▸ Time: 193,448
[15:35:47]: ▸ There was 1 failure:
[15:35:47]: ▸ 1) jPScreenshotTest(org.wordpress.android.ui.screenshots.JPScreenshotTest)
[15:35:47]: ▸ androidx.test.espresso.base.DefaultFailureHandler$AssertionFailedWithCauseError: 'an instance of android.widget.TextView and view.getText() with or without transformation to match: is "Choose site"' doesn't match the selected view.
[15:35:47]: ▸ Expected: an instance of android.widget.TextView and view.getText() with or without transformation to match: is "Choose site"
[15:35:47]: ▸      Got: view.getText() was "Choisir un site" transformed text was "Choisir un site"
[15:35:47]: ▸ View Details: AppCompatTextView{id=-1, visibility=VISIBLE, width=368, height=74, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.appcompat.widget.Toolbar$LayoutParams@ab38b50, tag=null, root-is-layout-requested=false, has-input-connection=false, x=198.0, y=40.0, text=Choisir un site, input-type=0, ime-target=false, has-links=false}
[15:35:47]: ▸ 	at dalvik.system.VMStack.getThreadStackTrace(Native Method)
[15:35:47]: ▸ 	at java.lang.Thread.getStackTrace(Thread.java:1538)
[15:35:47]: ▸ 	at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:12)
[15:35:47]: ▸ 	at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:7)
[15:35:47]: ▸ 	at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:5)
[15:35:47]: ▸ 	at androidx.test.espresso.ViewInteraction.check(ViewInteraction.java:12)
[15:35:47]: ▸ 	at org.wordpress.android.e2e.pages.MySitesPage.switchSite(MySitesPage.java:62)
[15:35:47]: ▸ 	at org.wordpress.android.e2e.pages.MySitesPage.startNewSite(MySitesPage.java:85)
[15:35:47]: ▸ 	at org.wordpress.android.ui.screenshots.JPScreenshotTest.generateSiteTopic(JPScreenshotTest.java:213)
[15:35:47]: ▸ 	at org.wordpress.android.ui.screenshots.JPScreenshotTest.jPScreenshotTest(JPScreenshotTest.java:135)
[15:35:47]: ▸ 	... 41 trimmed
[15:35:47]: ▸ Caused by: junit.framework.AssertionFailedError: 'an instance of android.widget.TextView and view.getText() with or without transformation to match: is "Choose site"' doesn't match the selected view.
[15:35:47]: ▸ Expected: an instance of android.widget.TextView and view.getText() with or without transformation to match: is "Choose site"
[15:35:47]: ▸      Got: view.getText() was "Choisir un site" transformed text was "Choisir un site"
[15:35:47]: ▸ View Details: AppCompatTextView{id=-1, visibility=VISIBLE, width=368, height=74, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.appcompat.widget.Toolbar$LayoutParams@ab38b50, tag=null, root-is-layout-requested=false, has-input-connection=false, x=198.0, y=40.0, text=Choisir un site, input-type=0, ime-target=false, has-links=false}
[15:35:47]: ▸ 	at androidx.test.espresso.matcher.ViewMatchers.assertThat(ViewMatchers.java:16)
[15:35:47]: ▸ 	at androidx.test.espresso.assertion.ViewAssertions$MatchesViewAssertion.check(ViewAssertions.java:11)
[15:35:47]: ▸ 	at androidx.test.espresso.ViewInteraction$SingleExecutionViewAssertion.check(ViewInteraction.java:2)
[15:35:47]: ▸ 	at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:12)
[15:35:47]: ▸ 	at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:1)
[15:35:47]: ▸ 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[15:35:47]: ▸ 	at android.os.Handler.handleCallback(Handler.java:873)
[15:35:47]: ▸ 	at android.os.Handler.dispatchMessage(Handler.java:99)
[15:35:47]: ▸ 	at android.os.Looper.loop(Looper.java:193)
[15:35:47]: ▸ 	at android.app.ActivityThread.main(ActivityThread.java:6669)
[15:35:47]: ▸ 	at java.lang.reflect.Method.invoke(Native Method)
[15:35:47]: ▸ 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
[15:35:47]: ▸ 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
[15:35:47]: ▸ FAILURES!!!
[15:35:47]: ▸ Tests run: 1,  Failures: 1
```

</details>

PS: I am just `commenting` on this PR, that is, without `approving` or `requesting changes`, since I am just being an extra pair of eyes on it. You will also find some additional questions (❓) and minor (🔍) comments from my side, I hope they will end-up being helpful.

AliSoftware commented:
Hello @ParaskP7 anf thanks a lot for testing this PR!

Could you indicate which configuration you tested? M1 vs Intel and which way you installed the Android command line tools?

---

As for the crash you reported when testing with Jetpack, thanks for that, but this is [a known one, already tracked here](https://github.com/wordpress-mobile/WordPress-Android/issues/17129). 🙃

As it is not related to the tooling itself but is instead a flaky issue with Hilt only on the `JPScreenshotTest.java` test and Android code itself, this will be investigated separately (likely by @AjeshRPai) later 🙂

ParaskP7 commented:
👋 @AliSoftware !

> Could you indicate which configuration you tested? M1 vs Intel and which way you installed the Android command line tools?

Sure, I tested with Mac on `Intel` chip and installed the Android command line via the AS (`SDK Manager`).

> As for the crash you reported when testing with Jetpack, thanks for that, but this is https://github.com/wordpress-mobile/WordPress-Android/issues/17129. 🙃

Great, yes, I was thinking that's the case, but wanted to note that anyway since in the testing instructions you are mentioning that we can also test with the relevant `Jetpack` fastlane command (then go for a long walk, etc.). I just wasn't sure... thank's for confirming that for me! 🙇 

> As it is not related to the tooling itself but is instead a flaky issue with Hilt only on the JPScreenshotTest.java test and Android code itself, this will be investigated separately (likely by @AjeshRPai) later 🙂

👍

AliSoftware committed:
Merge branch 'trunk' into tooling/screenshots

# Conflicts:
#	Gemfile.lock

AliSoftware committed:
Updating to latest changes from release-toolkit PR

AliSoftware committed:
Update to latest commit from release-toolkit branch

AliSoftware committed:
Reformat debug `AndroidManifest.xml`

As suggested in https://github.com/wordpress-mobile/WordPress-Android/pull/17096#discussion_r975312892

AliSoftware committed:
Fix extra semicolon in Kotlin code

As suggested in https://github.com/wordpress-mobile/WordPress-Android/pull/17096#discussion_r975314589

ovitrif committed:
Add info to remove confusion about AppInitializer.context non-null requirement

AliSoftware committed:
Merge pull request #17196 from wordpress-mobile/tooling/screenshots-updateContextLocale-improved-docs

Document updateContextLocale contract to remove confusion about context non-null requirement

AliSoftware committed:
Switch to latest release-toolkit

Now that 5.6.0 shipped, with all the new actions I need for this PR

AliSoftware commented:
Recalling the apps-infra team for review now that the release-toolkit 5.6.0 has been shipped and this PR can point to it — which should now make it ready to merge once approved. 🙏  🙇

AliSoftware committed:
Merge branch 'trunk' into tooling/screenshots

# Conflicts:
#	Gemfile.lock
#	WordPress/src/main/java/org/wordpress/android/WordPress.kt

AliSoftware committed:
Fix code change made in error (copy/paste)

See https://github.com/wordpress-mobile/WordPress-Android/pull/17096/files#r956389155

AliSoftware committed:
Remove the `create_emulators` lane

Now that each emulator is created on-demand at the moment the `screenshots` lane need them.
See https://github.com/wordpress-mobile/WordPress-Android/pull/17096/files#r956391257

AliSoftware committed:
Fix incorrect conflict resolution during backmerge

AliSoftware commented:
Just resync'd the PR with `trunk`, to fix conflicts that appeared as trunk continued to evolve since the month this PR has been waiting for review.

@wordpress-mobile/apps-infrastructure Can someone add their review please, so we can finally merge it before it diverges again? This is blocking my progress on the rest of my current project Thread 😢
